### ISSUE TITLE
Random identity card number

### ISSUE THREAD

**Issue by Shinigami92** [faker-js/faker#416]
### Clear and concise description of the problem

I would like to have a function to generate a random identity card number

In germany the implementation would follow this pattern: https://de.wikipedia.org/wiki/Ausweisnummer

Not sure right now how patterns in other country looks like, but in the long run I would need these also.

The sequence should be seed-based

### Suggested solution

Implement a function that returns an identity card number that is seed-based and related to the current selected locale

**Comment by import-brain** [comment]
What about locales that do not have a "national identity card"? The US doesn't have anything like that, would this function generate a Social Security number? Passport number?

**Comment by ST-DDT** [comment]
How about designing it more generic. We could use a Record<string, algorithm config> to define these things and the user has to know or lookup which generators are available for generateId(type: string).
The you can have one pattern for driving licences, for personal identity card numbers.

Or even more generically, a method that just takes the generation config and will build it for you.
You want a credit card with name, number, expiration date...? You get will get it.
You want an array of x-y foobars? You will get it.
You want a single identity card number that is generated based on a specific pattern? You will get it.
You just have to describe the thing you want.
Something like fake() and mustache() but for generic objects/arrays/elements.

**Comment by Shinigami92** [comment]
> How about designing it more generic. We could use a Record<string, algorithm config> to define these things and the user has to know or lookup which generators are available for generateId(type: string). The you can have one pattern for driving licences, for personal identity card numbers.
> 
> Or even more generically, a method that just takes the generation config and will build it for you. You want a credit card with name, number, expiration date...? You get will get it. You want an array of x-y foobars? You will get it. You want a single identity card number that is generated based on a specific pattern? You will get it. You just have to describe the thing you want. Something like fake() and mustache() but for generic objects/arrays/elements.

Yes, that would be nice in the long run, but I don't want to pass an algorithm or pattern as developer to faker, in that case I could just write it myself.
I would like to have it very basic and easy to use for now.
And if a locale doesn't have such thing, I'm satisfied with `undefined` for now.

**Comment by bryancross** [comment]
Algorithimically generated ID numbers like the _Ausweisnummer_ seem like a pretty heavy lift.  Seems like these should have their own defined functions, perhaps as part of a locale.  US Social Security numbers are easier, as they aren't algorithimically validated.  You can get a US SSN using the existing [`faker.helpers.regexpStyleStringParse`](https://github.com/faker-js/faker/blob/fda44bb899efbcbac2071a199964a3e269cbd805/src/modules/helpers/index.ts#L189)`, e.g.: 

`faker.helpers.regexpStyleStringParse('[0-9][0-9][0-9]-[0-9][0-9]-[0-9][0-9][0-9][0-9]')`

What might be cool is if this method could repeat string patterns as well as single characters, so that might condense down to: 

`faker.helpers.regexpStyleStringParse('[0-9](3)-[0-9](2)-[0-9](4)')l`

Or perhaps with common format identifiers, e.g.: 

`faker.helpers.regexpStyleStringParse('us-ssn');`

There are likely multiple "National ID"s for any given country.  The US has SSN, passport number, driver's license number (which varies by state), PASSID/KTN, license plates (again varying by state), etc.

**Comment by ST-DDT** [comment]
Since this varies greatly in both available ids and their implementation we have two choices:
- Move implementation to the locales.
- Implement it as a new (plugin-)module

I don't see this in v7 and even v8.

**Comment by ST-DDT** [comment]
If you are interested in this feature, please upvote this issue.

**Comment by guitartsword** [comment]
Maybe this will help create some random number SSN-like value:

```
faker.phone.number('###-##-####')
// Example Results:
// 771-00-9483
// 315-74-5571
```

For a "valid" Honduras ID number:

```
faker.phone.number('####-####-#####')
```

It creates something that may work, but may create some invalid values, for Honduras ID, the first 2 digits is the Department, the next two digits is the municipality, next 4 is year issued, next 5 are indeed random. The year is easy to generate, but the first 4 requires more development.

I like this idea:
> Implement it as a new (plugin-)module (per country)

**Comment by 0bvim** [comment]
In Brazil we `CPF`, `RG` and `PASSPORT` like `SSN` in US. Both with easy patterns to implement. It should be very helpful when testing user registration and so forth.

**Comment by Shinigami92** [comment]
**Team Decision**:

- The issue has collected enough interest, so we considering it
- We will work internally on an implementation-concept
- After the concept we will search for a champion in the community to implement the feature

_feedback from the community is appreciated_

**Comment by Marcos7765** [comment]
Hi, sorry to bother if this isn't the appropriate place. Where do i check for updates for it btw?

I've thought about using it as `Faker.person.document` and it either being a submodule whose entries should be given by the locale or a function that receives a `({ type: "<some_type_of_doc>" }, ...args)` and forwards args to whatever function matches the type (again, given by the locale). It does feel like it would explode the size of the Reference though (prolly a per-locale reference for these would be necessary?). I can try to help out on some of that if it means we get `CPF` and `CNH` generation for PT-BR

**Comment by matthewmayer** [comment]
I think it would be useful to make a draft list of which ID card numbers could be generated per locale that are feasible with just `faker.helpers.fake` patterns versus ones that might need custom code (eg if there is a checkdigit).
